sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/RedundantMethodsInDataClassesCheck.kt

```diff
@@ -0,0 +1,186 @@
+                context.reportIssue(it.nameIdentifier!!, issueMessage)
+            }
+        }
+    }
+
+    private fun KtNamedFunction.hashCodeHasDefaultImpl(klassParameters: List<KtParameter>, bindingContext: BindingContext): Boolean {
+        return if (hasBlockBody()) {
+            val returnExpressions = collectDescendantsOfType<KtReturnExpression>()
+            if (returnExpressions.size > 1) return false
+            checkHashExpression(returnExpressions[0], bindingContext, klassParameters)
+        } else {
+            checkHashExpression(this, bindingContext, klassParameters)
+        }
+    }
+
+    private fun checkHashExpression(expression: KtExpression, bindingContext: BindingContext, klassParameters: List<KtParameter>): Boolean {
+        val callExpressions = expression.collectDescendantsOfType<KtCallExpression> {
+            OBJECT_ARRAY_HASH_MATCHER.any { e ->
+                e.matches(it, bindingContext)
+            }
+        }
+        if (callExpressions.size > 1 || callExpressions.isEmpty()) return false
+
+        if ((callExpressions[0].calleeExpression as KtNameReferenceExpression).getReferencedName() == "hash") {
+            if (callExpressions[0].valueArguments.size != klassParameters.size) return false
+            callExpressions[0].valueArguments.forEach {
+                findParameter(it.getArgumentExpression(), klassParameters) ?: return false
+            }
+        } else {
+            val arguments = callExpressions[0].valueArguments[0].collectDescendantsOfType<KtValueArgumentList>()
+            if (arguments.isEmpty() || arguments.size > 1) return false
+
+            if (arguments[0].arguments.size != klassParameters.size) return false
+            arguments[0].arguments.forEach {
+                findParameter(it.getArgumentExpression(), klassParameters) ?: return false
+            }
+        }
+        return true
+    }
+
+    private fun KtNamedFunction.equalsHasDefaultImpl(
+            klassParameters: List<KtParameter>,
+            methodParameters: MutableList<KtParameter>,
+            className: String?
+    ): Boolean {
+        return if (hasBlockBody()) {
+            val returnExpressions = collectDescendantsOfType<KtReturnExpression>()
+            if (returnExpressions.size > 1) return false
+            checkEqualsExpression(returnExpressions[0], klassParameters, methodParameters, className)
+        } else {
+            val expression = findDescendantOfType<KtBinaryExpression>() ?: return false
+            checkEqualsExpression(expression, klassParameters, methodParameters, className)
+        }
+    }
+
+    private fun checkEqualsExpression(expression: KtExpression, klassParameters: List<KtParameter>, methodParameters: MutableList<KtParameter>, className: String?): Boolean {
+        val notEqualsExpressions = expression.collectDescendantsOfType<KtBinaryExpression> { it.operationToken == KtTokens.EXCLEQ }
```

Comment:
I tried to think of a better implementation, as I don't really like, when we use `.collectDescendantsOfType` as it ignores other expressions and we might face FPs here.

How about:
1. Check if expression is binary expression
2. recursively visit expression.right and expression.left
3. if one of them is not a KtBinaryExpression or a leaf, we return 

Knowing that the parser is left-recursive, leaves should be on the right

This way we can eliminate cases when there is something else between binary Expressions